### PR TITLE
GG-39863 erase empty cache directory after cache deletion

### PR DESCRIPTION
**Pull request by odiszapc**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java [hunk 1]
```diff
@@ -291,8 +292,6 @@ private long checkFile(FileIO fileIO, File cfgFile) throws IOException {
 
             fileIO = null;
 
-            if (delete)
-                Files.delete(pathProvider.apply());
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java:295]: Can you please explain why you removed these two lines?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java [hunk 2]
```diff
@@ -1093,8 +1095,13 @@ private IgniteCheckedException shutdown(Collection<CacheStoreHolder> holders, bo
      */
     private IgniteCheckedException shutdown(CacheStoreHolder holder, boolean cleanFile,
         @Nullable IgniteCheckedException aggr) {
+
         aggr = shutdown(holder.idxStore, cleanFile, aggr);
 
+        if (cleanFile) {
+            cleanParentDirectories(holder);
+        }
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1103]: What's the reason behind deleting a directory **before** you close page store instances for its content? This code looks broken, I don't understand your intentions

File: modules/core/src/test/java/org/apache/ignite/cache/CacheDirectoryCleanupTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.ignite.internal.processors.cache.persistence.file.FilePageStoreManager;
+import org.apache.ignite.testframework.junits.WithSystemProperty;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_DISABLE_WAL_DURING_REBALANCING;
+
+/** */
+@WithSystemProperty(key = IGNITE_DISABLE_WAL_DURING_REBALANCING, value = "false")
+public class CacheDirectoryCleanupTest extends GridCommonAbstractTest {
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/cache/CacheDirectoryCleanupTest.java:37]: Please add your test to a suite, it won't run on TC right now, I believe you didn't check it

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java [hunk 4]
```diff
@@ -1093,8 +1095,13 @@ private IgniteCheckedException shutdown(Collection<CacheStoreHolder> holders, bo
      */
     private IgniteCheckedException shutdown(CacheStoreHolder holder, boolean cleanFile,
         @Nullable IgniteCheckedException aggr) {
+
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1098]:
```suggestion
        @Nullable IgniteCheckedException aggr
    ) {
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java [hunk 5]
```diff
@@ -1103,6 +1110,23 @@ private IgniteCheckedException shutdown(CacheStoreHolder holder, boolean cleanFi
         return aggr;
     }
 
+    private void cleanParentDirectories(CacheStoreHolder holder) {
+        Arrays.stream(holder.partStores).map(store -> ((FilePageStoreV2)store).getPath().getParent()).distinct().forEach(new Consumer<Path>() {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1114]: Doing such class casts is a bad habit, maybe we should introduce new method to interface

Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1114]: All files in a store should have the same parent, it should be enough to read it for index partition, for example.
Or, if you want to be maximally flexible, you should use `findAny` instead of `distinct` stream.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java [hunk 6]
```diff
@@ -1103,6 +1110,23 @@ private IgniteCheckedException shutdown(CacheStoreHolder holder, boolean cleanFi
...
+            @Override public void accept(Path dir) {
+                if (Files.isDirectory(dir, LinkOption.NOFOLLOW_LINKS)) {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1116]: I believe you can reuse `org.apache.ignite.internal.util.IgniteUtils#delete(java.nio.file.Path)`, it is recursive

File: modules/core/src/test/java/org/apache/ignite/cache/CacheDirectoryCleanupTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.processors.cache.persistence.file.FilePageStoreManager;
+import org.apache.ignite.testframework.junits.WithSystemProperty;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_DISABLE_WAL_DURING_REBALANCING;
+
+/** */
+@WithSystemProperty(key = IGNITE_DISABLE_WAL_DURING_REBALANCING, value = "false")
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/cache/CacheDirectoryCleanupTest.java:36]: Why did you add this?

File: modules/core/src/test/java/org/apache/ignite/cache/CacheDirectoryCleanupTest.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        ignite1.createCache(new CacheConfiguration(DEFAULT_CACHE_NAME));
+        IgniteCache c = ignite1.cache(DEFAULT_CACHE_NAME);
+        c.put(30000, 30000);
+
+        c.destroy();
+
+        File workDir = ((FilePageStoreManager)ignite1.context().cache().context().pageStore()).workDir();
+        assertFalse(new File(workDir, "cache-" + DEFAULT_CACHE_NAME).exists());
+    }
+
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/cache/CacheDirectoryCleanupTest.java:86]:
```suggestion
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java (hunk 1 above)
Comment (odiszapc) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java]: mistake. Reverted

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java (hunk 2 above)
Comment (odiszapc) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java]: mistake. fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java [hunk 9]
```diff
@@ -1093,17 +1096,35 @@ private IgniteCheckedException shutdown(Collection<CacheStoreHolder> holders, bo
...
+            cleanParentDirectories(holder);
+
         return aggr;
     }
 
+    private void cleanParentDirectories(CacheStoreHolder holder) {
+        Arrays.stream(holder.partStores).map(store -> store.getPath().getParent()).findAny().ifPresent(dir -> {
+            if (Files.isDirectory(dir, LinkOption.NOFOLLOW_LINKS)) {
+                try {
+                    if (!Files.list(dir).findFirst().isPresent())
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1118]: I see, you only remove the directory if it's not empty. What happens to `cache_data.dat` file, do we remove it when we delete cache? I'm not sure I know where that happens.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java [hunk 10]
```diff
@@ -394,4 +396,18 @@ public void cacheDestroyWithConcImplicitTx() throws Exception {
...
+        IgniteEx ignite = startGrids(1);
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        ignite.createCache(new CacheConfiguration(DEFAULT_CACHE_NAME));
+        IgniteCache cache = ignite.cache(DEFAULT_CACHE_NAME);
+        cache.put(30000, 30000);
+
+        cache.destroy();
+
+        assertFalse(new File(((FilePageStoreManager)ignite.context().cache().context().pageStore()).workDir(), "cache-" + DEFAULT_CACHE_NAME).exists());
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java:411]: This line seems too long

Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java:411]: And just in case, please assert that this dir exists before destroying the cache, so that there's no doubt that there are no error in test itself

**Review by ibessonov — APPROVED** [review]
Please expand the test a little bit, and then everything's fine. Giving you an approval in advance

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java (hunk 9 above)
Comment (odiszapc) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java:1118]: Yes, because `cache_data.dat` is removed a little bit earlier as a part of entire cache decommission process here: https://github.com/gridgain/gridgain/blob/c942a4f899b5228a6c8084dfe58f34f9de6afdc7/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStoreManager.java#L1166-L1166

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java (hunk 10 above)
Comment (odiszapc) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsDestroyCacheTest.java]: added additional assert

**Review by ibessonov — APPROVED** [review]
